sdk/schemaregistry/Azure.Data.SchemaRegistry/src/Serialization/SchemaRegistrySerializer.cs

```diff
@@ -0,0 +1,550 @@
+        /// have a parameterless constructor.
+        /// If left blank, the serialized data will be set in a <see cref="MessageContent"/> instance.</param>
+        /// <param name="cancellationToken">An optional <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
+        /// <exception cref="InvalidOperationException">
+        ///   This can occur if the <c>groupName</c> was not specified when constructing the <see cref="SchemaRegistrySerializer"/>.
+        ///   It can also occur if the <paramref name="messageType"/> does not have a public parameterless constructor.
+        /// </exception>
+        /// <exception cref="RequestFailedException">
+        ///   An error occurred while attempting to communicate with the Schema Registry service. This can occur if the schema generated from the type
+        ///   does not exist in the registry.
+        /// </exception>
+        /// <exception cref="Exception">
+        ///   The data did not adhere to the specified schema, or the schema itself was invalid. The inner exception will hold more detailed information about
+        ///   the exception.
+        /// </exception>
+        public async ValueTask<MessageContent> SerializeAsync(
+            object data,
+            Type dataType = default,
+            Type messageType = default,
+            CancellationToken cancellationToken = default)
+            => await SerializeInternalAsync(data, dataType, messageType, true, cancellationToken).ConfigureAwait(false);
+
+        internal async ValueTask<MessageContent> SerializeInternalAsync(
+            object data,
+            Type dataType,
+            Type messageType,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            if (_groupName == null)
+            {
+                throw new InvalidOperationException(
+                    "A group name must be specified in the 'SchemaRegistryJsonSerializer' constructor if you will be attempting to serialize. " +
+                    "The group name can be omitted if only deserializing.");
+            }
+
+            messageType ??= typeof(MessageContent);
+            if (messageType.GetConstructor(Type.EmptyTypes) == null)
+            {
+                throw new InvalidOperationException(
+                    $"The type {messageType} must have a public parameterless constructor in order to use it as the 'MessageContent' type to serialize to.");
+            }
+
+            var message = (MessageContent)Activator.CreateInstance(messageType);
+
+            (string retrievedSchemaId, BinaryData bd) = async
+                ? await SerializeInternalAsync(data, dataType, true, cancellationToken).ConfigureAwait(false)
+                : SerializeInternalAsync(data, dataType, false, cancellationToken).EnsureCompleted();
+
+            message.Data = bd;
+            message.ContentType = $"{_mimeType}+{retrievedSchemaId}";
```

Comment:
Isn't this string also LRU cache-able?

Reply:
Yeah that's true we could also do a mapping from contentType --> schema, rather than schemaId --> schema. Since we would need the schemaId to interact with the service if it's not cached, I think it felt more natural to go with the schemaId --> schema mapping.

Reply:
I was mostly thinking from the angle of having to do a string concatenation on the hot path that could be avoided potentially